main.tf

```diff
@@ -17,35 +28,97 @@ resource "alicloud_vpc" "vpc" {
+# the Internet from initiating a connection to those instances.
+# See https://www.alibabacloud.com/product/nat
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "alicloud_nat_gateway" "this" {
+  count         = var.create_nat_gateway ? 1 : 0
+  vpc_id        = local.vpc_id
+  name          = var.vpc_name
+  specification = var.nat_gateway_specification
+}
+
+# create one or more eips for NAT GW
+resource "alicloud_eip" "nat" {
+  name      = var.vpc_name
+  count     = var.create_nat_gateway && var.nat_gateway_num_eips > 0 ? var.nat_gateway_num_eips : 0
+  bandwidth = var.nat_eip_bandwidth
+  tags      = var.tags
+}
+
+# attach eips to nat gateway
+resource "alicloud_eip_association" "nat" {
+  count         = var.create_nat_gateway && var.nat_gateway_num_eips > 0 ? var.nat_gateway_num_eips : 0
+  allocation_id = alicloud_eip.nat[count.index].id
+  instance_id   = alicloud_nat_gateway.this[0].id
+}
+
+# ---------------------------------------------------------------------------------------------------------------------
+#  VSwitch
+# ---------------------------------------------------------------------------------------------------------------------
+resource "alicloud_vswitch" "this" {
+  for_each          = var.vswitches
+  vpc_id            = local.vpc_id
+  name              = each.value.name
+  availability_zone = each.value.availability_zone
+  cidr_block        = each.value.cidr_block
+  description       = each.value.description
+
   tags = merge(
     {
       Name = format(
-        "%s%03d",
-        var.vswitch_name,
-        count.index + 1
+        "%s",
+        "${each.value.name}"
       )
     },
-    var.vswitch_tags,
+    var.tags,
+    var.vswitches_tags,
   )
+
 }
 
-// According to the destination cidr block to launch a new route entry
-resource "alicloud_route_entry" "route_entry" {
-  count                 = local.create_sub_resources ? length(var.destination_cidrs) : 0
-  route_table_id        = local.route_table_id
-  destination_cidrblock = var.destination_cidrs[count.index]
-  nexthop_type          = "Instance"
-  nexthop_id            = var.nexthop_ids[count.index]
+# Create a dedicated Route Table for vswitches
+# Alibaba has no notion of Internet Gateway, it may be necessary to add custom routes for certain VSwitches
+resource "alicloud_route_table" "this" {
+  count       = length(var.vswitches) > 0 ? 1 : 0
+  vpc_id      = local.vpc_id
+  name        = var.route_table_name
+  description = var.route_table_description
+  tags        = var.tags
 }
+
+# Associate each vswitch with the route table
+resource "alicloud_route_table_attachment" "this" {
+  for_each       = alicloud_vswitch.this
+  vswitch_id     = each.value.id
+  route_table_id = alicloud_route_table.this.0.id
+}
+
+resource "alicloud_route_entry" "this" {
+  for_each              = var.custom_routes
+  route_table_id        = alicloud_route_table.this[0].id
```

Comment:
done

Reply:
I would like to understand what is the purpose to do this. If the route table is not built, it will throw an error. I have added the option to specify a different route table as it might be useful